storage,kv: always consider intents in RangeLookup

tamird opened the pull request:
Following up from #9359.

Nothing really interesting here, since we're almost always considering intents anyway.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9416)

<!-- Reviewable:end -->

tamird committed:
kv: move test-only function to test

tamird committed:
storage,kv: always consider intents in RangeLookup

tschottdorf commented:
The ReplicaGCQueue tries to do an authoritative lookup of the RangeDescriptor when it makes its GC decision. Will this change impact that decision making process and if so, how?

---

Reviewed 2 of 2 files at r1.
Review status: 0 of 13 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-:-KRndqb1fnYMVI1cOdg7:b-idex08)*
<!-- Sent from Reviewable.io -->

tamird commented:
The replicaGCQueue does a consistent lookup, which will force intents to be
resolved, rather than be returned. Thus, the behavior is not affected by
this change.

Have I misunderstood you?

On Sep 16, 2016 11:41, "Tobias Schottdorf" <notifications@github.com> wrote:

> The ReplicaGCQueue tries to do an authoritative lookup of the
> RangeDescriptor when it makes its GC decision. Will this change impact that
> decision making process and if so, how?
> ------------------------------
>
> Reviewed 2 of 2 files at r1.
> Review status: 0 of 13 files reviewed at latest revision, all discussions
> resolved, all commit checks successful.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-:-KRndqb1fnYMVI1cOdg7:b-idex08>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9416#issuecomment-247633923>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPE4XjR2S9yrISNvD8yUT4GQ2rewfks5qqriOgaJpZM4J-YHb>
> .
>

tschottdorf commented:
Ah, right. Thanks for the reminder.

On Fri, Sep 16, 2016 at 11:58 AM Tamir Duberstein <notifications@github.com>
wrote:

> The replicaGCQueue does a consistent lookup, which will force intents to be
> resolved, rather than be returned. Thus, the behavior is not affected by
> this change.
>
> Have I misunderstood you?
>
> On Sep 16, 2016 11:41, "Tobias Schottdorf" <notifications@github.com>
> wrote:
>
> > The ReplicaGCQueue tries to do an authoritative lookup of the
> > RangeDescriptor when it makes its GC decision. Will this change impact
> that
> > decision making process and if so, how?
> > ------------------------------
> >
> > Reviewed 2 of 2 files at r1.
> > Review status: 0 of 13 files reviewed at latest revision, all discussions
> > resolved, all commit checks successful.
> > ------------------------------
> >
> > *Comments from Reviewable
> > <
> https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-:-KRndqb1fnYMVI1cOdg7:b-idex08
> >*
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/cockroachdb/cockroach/pull/9416#issuecomment-247633923
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/ABdsPE4XjR2S9yrISNvD8yUT4GQ2rewfks5qqriOgaJpZM4J-YHb
> >
> > .
> >
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9416#issuecomment-247638279>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135B8YEnc9bYO9g6ISJ-yHjl9RsB7mks5qqryXgaJpZM4J-YHb>
> .
>
-- 

-- Tobias

andreimatei commented:
LGTM

Can this code go away, given https://github.com/cockroachdb/cockroach/commit/b6af8967f568143512e202a3587928f693b7b3a3?
https://github.com/cockroachdb/cockroach/blob/06b04672355abac49bf701781fd583e46b70467e/kv/dist_sender.go#L712
To be honest I don't understand the case this check is protecting against. And the comment seems malformed to me.
Also, the range cache seems to be also validating results  (https://github.com/cockroachdb/cockroach/blob/06b0467/kv/range_cache.go#L407), so I don't understand why this code could be necessary. And also I don't understand why the bug you previously fixed would ever make it to the client. What am I missing?

---

Review status: 0 of 13 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[kv/dist_sender.go, line 427 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRny2bcgXq6E7QG9Ya7:-KRny2bcgXq6E7QG9Ya8:b-wc9tpl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/kv/dist_sender.go#L427)):*
> ```Go
> 	desc, returnToken, err := ds.rangeCache.LookupRangeDescriptor(
> 		ctx,
> 		descKey,
> ```

nit: put them on the same line

---

*[kv/range_cache.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo10MZWC2MRwt8Jw3x:-KRo10MZWC2MRwt8Jw3y:bcfxwge) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b6afeb7e17956e788320579ca3b96de760aa94/kv/range_cache.go#L63)):*
> ```Go
> 	// the given key (possibly from intents), and the second being prefetched
> 	// adjacent descriptors.
> 	// TODO(andrei): Should this return error instead of pErr?
> ```

I think this TODO is valid; this should return `error`.

---

*[kv/range_cache.go, line 57 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo1AYEaQMWUaxyn1JM:-KRo1AYEaQMWUaxyn1JN:b-jttsx7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/kv/range_cache.go#L57)):*
> ```Go
> // initially retrieve information which will be cached.
> type RangeDescriptorDB interface {
> 	// RangeLookup takes a meta key to look up descriptors for and a descriptor
> ```

say that it can either be a meta1 or meta2 key?

---

*[roachpb/api.proto, line 384 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo1QrhD8HXN3ou73eF:-KRo1QrhD8HXN3ou73eG:bd5k12o) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/roachpb/api.proto#L384)):*
> ```protobuf
>   optional Span header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
>   optional int32 max_ranges = 2 [(gogoproto.nullable) = false];
>   reserved 3; // previously consider_intents, now always true.
> ```

I think this kind of historical comment doesn't help. Are we gonna remember what `consider_intents` meant, and what `true` did?

---

*[storage/client_replica_test.go, line 449 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo1hpldjIFYVXYIKJl:-KRo1hpm5Uy62k3lGPyi:b-k0zb6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/storage/client_replica_test.go#L449)):*
> ```Go
> 		rlReply := resp.(*roachpb.RangeLookupResponse)
> 		// Checks the results count.
> 		if rsCount, preRSCount := len(rlReply.Ranges), len(rlReply.PrefetchedRanges); int32(rsCount+preRSCount) != test.request.MaxRanges {
> ```

nit: split on assignments to a separate statement

---

*[storage/replica_command.go, line 948 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo2Hutd4YiVHMfoQX2:-KRo2Hutd4YiVHMfoQX3:bcdqz75) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/storage/replica_command.go#L948)):*
> ```Go
> 		return reply, nil, errors.Errorf("illegal lookup of range-local key %q", args.Key)
> 	}
> 	ts, txn, consistent, rangeCount := h.Timestamp, h.Txn, h.ReadConsistency != roachpb.INCONSISTENT, int64(args.MaxRanges)
> ```

nits: why not use the members of `h` everywhere?
- `consistent`, if kept, I think would better be defined as `h.ReadConsistency != roachpb.CONSISTENT`
- why 4 assignments on one line? Hard to read, particularly since they're not all exact copies from `h`

---

*[storage/replica_command.go, line 1125 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo32ax-yontQTWeSXD:-KRo32ax-yontQTWeSXE:b-k995mq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/storage/replica_command.go#L1125)):*
> ```Go
> 		}
> 
> 		if val != nil {
> ```

nit: I liked it more before because it explained what `val==nil` means. And also no need to pollute the file history.

---

*[storage/replica_command.go, line 1144 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo3XPpeFk_985oA8vH:-KRo3XPpeFk_985oA8vI:b-54dabk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3128eba182097c600397f3cb1da70c3a8cece4a5/storage/replica_command.go#L1144)):*
> ```Go
> 	}
> 
> 	for _, rd := range reply.Ranges {
> ```

how come this went away?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-:-KRo-RbLp_5C-oT2yXPh:b-civ59p)*
<!-- Sent from Reviewable.io -->

tamird commented:
I don't know about the check in dist sender - I'll have to look closer tomorrow, but the check in range cache is necessary; from my conversation with @nvanbenschoten about it:
> tamird [18:58]  
anyway, the real question: now that this is handled in RangeLookup, does it makes sense to remove the handling in range cache?

> nathan [18:58]  
no we still need that specifically for two requests to different keys that are coalesced onto the same request

> [18:59]  
Where only one of them is within the new range descriptor

> tamird [18:59]  
can you elborate?

> nathan [19:00]  
so request for ‘a’ and ‘b’ are believed to be to the same descriptor, so we coalesce the requests. the returned descriptor is ‘a’ - ‘az’. ‘b’ needs to perform a new lookup but ‘a’ doesnt.

> tamird [19:01]  
“believed” because that’s what the cached descriptor indicates?

> nathan [19:01]  
because they were both part of the same range before a split

> tamird [19:01]  
according to the cached descriptor

> nathan [19:01]  
correct

> tamird [19:01]  
ok, cool, makes sense

Regarding the bug I previously fixed: it would make it to the client if the client was not the dist sender (e.g. replicaGCQueue).

---

Review status: 0 of 13 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[kv/dist_sender.go, line 427 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRny2bcgXq6E7QG9Ya7:-KRr86D2d6GsvzMjai8Q:btl4mh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/kv/dist_sender.go#L427)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> nit: put them on the same line
</details>
I deliberately did it this way. Is there a precedent for the formatting you're suggesting?

---

*[kv/range_cache.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo10MZWC2MRwt8Jw3x:-KRqsmDotlkh832vxVqx:b-4udkmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b6afeb7e17956e788320579ca3b96de760aa94/kv/range_cache.go#L63)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I think this TODO is valid; this should return `error`.
</details>
It can't, because the batch error is a `*roachpb.Error`. I would rather that this low level API remain lossless.

---

*[kv/range_cache.go, line 57 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo1AYEaQMWUaxyn1JM:-KRqsjjei1rJOpt0Qoab:bptwwsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/kv/range_cache.go#L57)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> say that it can either be a meta1 or meta2 key?
</details>
As opposed to other metas?

---

*[roachpb/api.proto, line 384 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo1QrhD8HXN3ou73eF:-KRqsvq9ktkPe4PnraEK:bdofmq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/roachpb/api.proto#L384)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I think this kind of historical comment doesn't help. Are we gonna remember what `consider_intents` meant, and what `true` did?
</details>
Would you rather I removed it? Happy to.

---

*[storage/client_replica_test.go, line 449 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo1hpldjIFYVXYIKJl:-KRqsyfpTZOZR2ijRrbi:bmvdo39) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/storage/client_replica_test.go#L449)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> nit: split on assignments to a separate statement
</details>
why? what's the precedent or rule?

---

*[storage/replica_command.go, line 948 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo2Hutd4YiVHMfoQX2:-KRqt6yntr_lOfl-cnxx:b1b1h6s) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/storage/replica_command.go#L948)):*
> nits: why not use the members of h everywhere?

No real reason; just following original precedent and 
 
> consistent, if kept, I think would better be defined as h.ReadConsistency != roachpb.CONSISTENT
> 
> why 4 assignments on one line? Hard to read, particularly since they're not all exact copies from h

---

*[storage/replica_command.go, line 1125 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo32ax-yontQTWeSXD:-KRr8AsMu4xLdFLjEGkW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/storage/replica_command.go#L1125)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> nit: I liked it more before because it explained what `val==nil` means. And also no need to pollute the file history.
</details>
Done.

---

*[storage/replica_command.go, line 1144 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo3XPpeFk_985oA8vH:-KRr7tWRGH_kEo29HGra:b-1uwgys) ([raw file](https://github.com/cockroachdb/cockroach/blob/3128eba182097c600397f3cb1da70c3a8cece4a5/storage/replica_command.go#L1144)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> how come this went away?
</details>
It's impossible to hit this condition. descriptors are only added to `.Ranges` if `containsFn(rd, userKey)`. We'll hit the above assertion on `len(reply.Ranges) == 0` in case of a gap in meta2.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-:-KRr8mYIrR4uNNuTdMWq:b-2z3ho9)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 13 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/replica_command.go, line 948 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo2Hutd4YiVHMfoQX2:-KRrl2CwG1qhnDdXjiu5:b6d0n7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/storage/replica_command.go#L948)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> > nits: why not use the members of h everywhere?
> 
> No real reason; just following original precedent and 
>  
> > consistent, if kept, I think would better be defined as h.ReadConsistency != roachpb.CONSISTENT
> > 
> > why 4 assignments on one line? Hard to read, particularly since they're not all exact copies from h
> 
</details>
oops, partial reply. following precedent and the variable names are shorter, making some lines below shorter.

I'll break the assignments into multiple lines, if you prefer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 13 of 13 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[kv/dist_sender.go, line 427 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRny2bcgXq6E7QG9Ya7:-KRwct8T6W4RPwBEHa5w:b-asa1dx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/kv/dist_sender.go#L427)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I deliberately did it this way. Is there a precedent for the formatting you're suggesting?
</details>
The line itself is the precedent; this pattern is pretty common (or was, before someone decided to get prescriptive about this stuff in the style guide). 

---

*[storage/client_replica_test.go, line 449 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo1hpldjIFYVXYIKJl:-KRwbWiW5fKMXACK5nwA:b-aq0b2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/storage/client_replica_test.go#L449)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why? what's the precedent or rule?
</details>
The 100-column line limit, for one thing.

The general rule (which I've seen in many style guides but not the go CodeReviewComments) is to do one statement per line; you wouldn't normally use semicolons to put multiple statements on one line (even though `k1 := v1; k2 := v2` is more legible IMHO than `k1, k2 := v1, v2` when you don't have the syntactic constraints of an `if` statements). 

That said, there is a reason to put the two assignments together here because of the scoping rules, and there's precedent for multi-assignments in `if` statements in tests, so I don't mind it here. (I would split up the other multi-assignment called out in this PR, though)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-:-KRwd1OUqtr8bppkjg1o:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 13 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[kv/dist_sender.go, line 427 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRny2bcgXq6E7QG9Ya7:-KS75avodo1IkJAgcqeb:bboi9v0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/kv/dist_sender.go#L427)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The line itself is the precedent; this pattern is pretty common (or was, before someone decided to get prescriptive about this stuff in the style guide). 
</details>
Done throughout.

---

*[storage/client_replica_test.go, line 449 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo1hpldjIFYVXYIKJl:-KS74fQnXaFGX5QOZlwD:b-a25g4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/storage/client_replica_test.go#L449)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The 100-column line limit, for one thing.
> 
> The general rule (which I've seen in many style guides but not the go CodeReviewComments) is to do one statement per line; you wouldn't normally use semicolons to put multiple statements on one line (even though `k1 := v1; k2 := v2` is more legible IMHO than `k1, k2 := v1, v2` when you don't have the syntactic constraints of an `if` statements). 
> 
> That said, there is a reason to put the two assignments together here because of the scoping rules, and there's precedent for multi-assignments in `if` statements in tests, so I don't mind it here. (I would split up the other multi-assignment called out in this PR, though)
</details>
Ack. Leaving this one as as-is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416)*
<!-- Sent from Reviewable.io -->

tamird committed:
kv: avoid allocating a closure

tamird committed:
kv: simplify

tamird committed:
kv: minor comment cleanup

tamird committed:
roachpb: update RangeKeyMismatchError comments

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 13 of 13 files at r5, 2 of 2 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, 2 of 2 files at r9, 13 of 13 files at r10.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[kv/range_cache.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo10MZWC2MRwt8Jw3x:-KS87TjUdeaeh3tfdJEw:b40wcud) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b6afeb7e17956e788320579ca3b96de760aa94/kv/range_cache.go#L63)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> It can't, because the batch error is a `*roachpb.Error`. I would rather that this low level API remain lossless.
</details>
well this layer doesn't need to be lossless. Remember, we moved away from roachpb.Error in all places whose errors does not need to be put over the wire. And that was a good thing :)

---

*[kv/range_cache.go, line 57 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo1AYEaQMWUaxyn1JM:-KS877S9Ei4TKVKYkv9e:b-827slq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/kv/range_cache.go#L57)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> As opposed to other metas?
</details>
The point is that it can be either one. At least for me that info is useful. If we just say "meta" to mean either in other places, then it's cool.

---

*[roachpb/api.proto, line 384 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo1QrhD8HXN3ou73eF:-KS87rPbXwDvkqW1w6Hj:b-fgn475) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/roachpb/api.proto#L384)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Would you rather I removed it? Happy to.
</details>
that's certainly what I would do :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-:-KS8Dhu6396-jw_g8fqT:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[kv/range_cache.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo10MZWC2MRwt8Jw3x:-KS8KNZAnYR3LPypCu9s:b-21b8a1) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b6afeb7e17956e788320579ca3b96de760aa94/kv/range_cache.go#L63)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> well this layer doesn't need to be lossless. Remember, we moved away from roachpb.Error in all places whose errors does not need to be put over the wire. And that was a good thing :)
</details>
In this case, a structured error may actually be useful to the caller (which is the other case, other than going over the wire, where `*roachpb.Error` should be used).

---

*[roachpb/api.proto, line 384 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo1QrhD8HXN3ou73eF:-KS8KgywSVfQE_eEO1US:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1191a32c3b9f7f452b06ab280a33597cffcdd30e/roachpb/api.proto#L384)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> that's certainly what I would do :)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 14 of 15 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[kv/range_cache.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo10MZWC2MRwt8Jw3x:-KSNWyEhCBD23HJJx8Z3:b-rga2ic) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b6afeb7e17956e788320579ca3b96de760aa94/kv/range_cache.go#L63)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> In this case, a structured error may actually be useful to the caller (which is the other case, other than going over the wire, where `*roachpb.Error` should be used).
</details>
errors can be just as "structured" without being protos. In this particular case, this guy is turned into an `error` immediately. Mind if I bring it back? Or I'll just do the thing.
If you look at the history, when the TODO was added Ben also agreed, so the tone of the TODO should be more authoritative.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 14 of 15 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[kv/range_cache.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo10MZWC2MRwt8Jw3x:-KSNYm7HCPn2-Vn5Xzqd:b9wzo6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b6afeb7e17956e788320579ca3b96de760aa94/kv/range_cache.go#L63)):*
> errors can be just as "structured" without being protos

Good point, though it will still be a proto, just not roachpb.Error. Do what you think is best.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 14 of 15 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[kv/range_cache.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416#-KRo10MZWC2MRwt8Jw3x:-KSN_1CarCsRVDAVfL-M:b-vvwutg) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b6afeb7e17956e788320579ca3b96de760aa94/kv/range_cache.go#L63)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> > errors can be just as "structured" without being protos
> 
> Good point, though it will still be a proto, just not roachpb.Error. Do what you think is best.
</details>
yeah, what I meant to say here is that they can be structured while implementing `error`. In particular, like the `ErrorDetail` protos do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9416)*
<!-- Sent from Reviewable.io -->
